### PR TITLE
Add a custom logging function - allows logging to file

### PR DESCRIPTION
**Pull request by RayyanAnsari**
This PR creates a custom logging function and gives an option to either log to console, or to log to file.
### Changes

- All `printf` calls are now `Platform::LogMessage`
- A new settings dialog has been created: Log Settings
- Two new variables are added to the config file: `LogToFile` and `LogFileLocation`
- The new `Platform::LogMessage` function uses either `vprintf` or `vfprintf` depending on the user's setting

### CODE REVIEW

File: src/frontend/qt_sdl/LogSettingsDialog.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+                                                "Select Log File...",
+                                                EmuDirectory,
+                                                "Log files (*.log *.txt);;Any file (*.*)");
+
+    if (file.isEmpty()) return;
+
+    ui->txtLogFolderPath->setText(file);
+}
+
+void LogSettingsDialog::updateAdapterControls()
```
Comment (RSDuck) [inline src/frontend/qt_sdl/LogSettingsDialog.cpp:87]: the name seems to be a left over

File: src/frontend/qt_sdl/Platform.cpp [hunk 2]
```diff
@@ -127,6 +131,24 @@ void StopEmu()
...
 
+void LogMessage(const char* format, ...)
+{
+    va_list args;
+    va_start(args, format);
+    if (Config::LogToFile)
+    {
+        if (!LogFile)
+        {
+            LogFile = fopen(Config::LogFileLocation, "a+");
```
Comment (RSDuck) [inline src/frontend/qt_sdl/Platform.cpp:142]: a+ seems to be dangerous, because melonDS can spit out a lot of data

Comment (WaluigiWare64) [reply src/frontend/qt_sdl/Platform.cpp:142]: should i change to w+?

File: src/frontend/qt_sdl/LogSettingsDialog.cpp (hunk 1 above)
Comment (WaluigiWare64) [reply src/frontend/qt_sdl/LogSettingsDialog.cpp]: ah yes, I copied it over from WifiSettingsDialog.cpp.

File: src/frontend/qt_sdl/Platform.cpp (hunk 2 above)
Comment (RSDuck) [reply src/frontend/qt_sdl/Platform.cpp:142]: only w we don't need to read

Comment (WaluigiWare64) [reply src/frontend/qt_sdl/Platform.cpp]: done

File: src/frontend/qt_sdl/Platform.cpp [hunk 3]
```diff
@@ -127,6 +131,24 @@ void StopEmu()
...
+        {
+            LogFile = fopen(Config::LogFileLocation, "w");
+        }
+        vfprintf(LogFile, format, args);
+    }
+    else
+    {
+        vprintf(format, args);       
+    }
+     va_end(args);
```
Comment (RSDuck) [inline src/frontend/qt_sdl/Platform.cpp:150]: va_end isn't properly aligned

Comment (WaluigiWare64) [reply src/frontend/qt_sdl/Platform.cpp]: done.

File: src/sha1/sha1.c [hunk 4]
```diff
@@ -267,10 +267,10 @@ main(int argc, char **argv)
         SHA1Update(&ctx, buf, BUFSIZE);
     SHA1Final(hash, &ctx);
 
-    Platform::LogMessage("SHA1=");
```
Comment (RSDuck) [inline src/sha1/sha1.c]: better not poke in external dependencies. Those are tests anyway, they are never run.

Comment (WaluigiWare64) [reply src/sha1/sha1.c]: Yep, which is why I changed it back to printf.

**Comment by nadiaholmquist** [comment]
If we're doing a custom logging thing, we really should also have log levels I think?

**Comment by WaluigiWare64** [comment]
We could, but currently this is just a wrapper around printf and fprintf. That would require modifying all of the printf statements we have.

**Comment by RayyanAnsari** [comment]
superseded by #1642

**Pull request closed without merging**